### PR TITLE
Ported CA2227

### PR DESCRIPTION
**Pull request by heejaechang**
ported CA2227 and added initial set of Microsoft.ApiDesignGuideLines.Analyzers

### CODE REVIEW

File: src/FxCop/Microsoft.ApiDesignGuidelines.Analyzers/Core/Usage/CollectionPropertiesShouldBeReadOnly.cs [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+                {
+                    var iCollectionType = context.Compilation.GetICollectionType();
+                    var arrayType = context.Compilation.GetArrayType();
+
+                    if (iCollectionType == null || arrayType == null)
+                    {
+                        return;
+                    }
+
+                    var analyzer = new Analyzer(iCollectionType, arrayType);
```
Comment (srivatsn) [inline src/FxCop/Microsoft.ApiDesignGuidelines.Analyzers/Core/Usage/CollectionPropertiesShouldBeReadOnly.cs]: Do you prefer explicitly creating the class here? You could simply have a lambda that captures these two variables and the compiler will generate the class. In general, I make the class explicit only when there's a lot of state being tracked and we are doing multiple things with that object. I'd suggest writing this as 
```csharp
context.RegisterSymbolAction(symbolContext => AnalyzeSymbol(iCollectionType, arrayType), SymbolKind.Property);
```

and make `AnalyzeSymbol` a static method on this class.

File: src/FxCop/Microsoft.ApiDesignGuidelines.Analyzers/Test/Usage/CollectionPropertiesShouldBeReadOnlyTests.cs [hunk 2]
```diff
@@ -0,0 +1,846 @@
...
+
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Test.Utilities;
+using Microsoft.CodeAnalysis.UnitTests;
+using Roslyn.Test.Utilities;
+using Xunit;
+
+namespace System.Runtime.Analyzers.UnitTests
+{
+    public partial class DisposableFieldsShouldBeDisposedTests : DiagnosticAnalyzerTestBase
```
Comment (srivatsn) [inline src/FxCop/Microsoft.ApiDesignGuidelines.Analyzers/Test/Usage/CollectionPropertiesShouldBeReadOnlyTests.cs]: What happened here? These are tests for a different rule

**Comment by genlu** [comment]
Could you please rebase and create a PR against `portFxCop` branch? This analyzer already exists there and would make merge easier in the future. Thanks!

File: src/FxCop/Microsoft.ApiDesignGuidelines.Analyzers/Test/Usage/CollectionPropertiesShouldBeReadOnlyTests.cs (hunk 2 above)
Comment (heejaechang) [inline src/FxCop/Microsoft.ApiDesignGuidelines.Analyzers/Test/Usage/CollectionPropertiesShouldBeReadOnlyTests.cs]: Not done yet ☺


-          heejae

**Comment by heejaechang** [comment]
@genlu thank you Gen. looks like all skeleton is there, I just need to port implementation! :)

**Comment by genlu** [comment]
@heejaechang I'm done creating the new branch. It is now called `FxCop`, all the newly generated code stubs are under `$root\FxCop` folder. Please create a PR against that branch. Thanks!

**Pull request closed without merging**